### PR TITLE
chore: minor fixes for v1.1.5

### LINKED ISSUE #1492: verify if the de-duplication logic is working as expected
**Describe the bug**
We need to verify the functionality of the Dedupe logic 
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

### LINKED ISSUE #1493: Reduce the number of Control Messages Published in pipelines with high partitions
We're observing a high number of control messages in multi-partition pipelines, which is affecting efficiency.

### PR DESCRIPTION
**Pull request by yhl25**
* fix dedup for reduce
 * fix seg fault when watermark is disabled in the pipeline
 * adding ISB latency metrics for better debugging in case of ISB issue
 * adding a metric to track the number of control messages

fixes #1493 
fixes #1492 
TODO(@jy4096): reduce the number of ctrl messages that are getting published in multi-partitioned pipeline

### CODE REVIEW

File: pkg/shared/clients/nats/nats_client.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func NewNATSClient(ctx context.Context, natsOptions ...nats.Option) (*NATSClient
 		nats.MaxReconnects(-1),
 		// every one second we will try to ping the server, if we don't get a pong back
 		// after two attempts, we will consider the connection lost and try to reconnect
-		nats.PingInterval(1 * time.Second),
+		nats.PingInterval(20 * time.Second),
```
Comment (whynowy) [inline pkg/shared/clients/nats/nats_client.go:55]: Is 20s too long?

File: pkg/udf/map_udf.go [hunk 2]
```diff
@@ -113,10 +117,6 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
 			// create watermark publisher using watermark stores
 			publishWatermark = jetstream.BuildPublishersFromStores(ctx, u.VertexInstance, toVertexWmStores)
 
-			readers, writers, err = buildJetStreamBufferIO(ctx, u.VertexInstance, natsClientPool)
```
Comment (whynowy) [inline pkg/udf/map_udf.go:116]: When watermark is disabled, it's not working? How come we have this bug.....

Comment (yhl25) [reply pkg/udf/map_udf.go:116]: it got introduced during a refactor

Comment (yhl25) [reply pkg/udf/map_udf.go:116]: updated one of the e2e tests to cover this case..

**Comment by vigith** [comment]
@yhl25 are you tracking the fixes so it can get to `main` too?

**Comment by yhl25** [comment]
> @yhl25 are you tracking the fixes so it can get to `main` too?

yes

**Comment by yhl25** [comment]
@jy4096 let's have a separate pr for fixing control messages.

**Comment by jy4096** [comment]
> @jy4096 let's have a separate pr for fixing control messages.

I've opened one PR based on main branch. Do we need one for release 1.1 as well?

File: pkg/shared/clients/nats/nats_client.go (hunk 1 above)
Comment (whynowy) [reply pkg/shared/clients/nats/nats_client.go:55]: We didn't change this in the PR to main.

File: pkg/shared/idlehandler/idlehandler.go [hunk 3]
```diff
@@ -72,3 +76,7 @@ func PublishIdleWatermark(ctx context.Context, toBufferPartition isb.BufferWrite
 		wmPublisher.PublishIdleWatermark(wm, nil, toVertexPartition)
 	}
 }
+
+// disable writing control messages and check if we still see duplicate messages
```
Comment (whynowy) [inline pkg/shared/idlehandler/idlehandler.go:80]: Something forgot to clean up?

File: pkg/udf/rpc/grpc_reduce.go [hunk 4]
```diff
@@ -155,6 +163,7 @@ func (u *GRPCBasedReduce) ApplyReduce(ctx context.Context, partitionID *partitio
 								IsLate:    false,
 							},
 							Keys: keys,
+							ID:   fmt.Sprintf("%s-%d-%s-%d", u.vertexName, u.vertexReplica, partitionID.String(), index),
```
Comment (whynowy) [inline pkg/udf/rpc/grpc_reduce.go:166]: Do we have this change to main?

**Comment by yhl25** [comment]
> > @jy4096 let's have a separate pr for fixing control messages.
> 
> I've opened one PR based on main branch. Do we need one for release 1.1 as well?

not required, we can have the fix in our 1.2 release.

File: pkg/shared/clients/nats/nats_client.go (hunk 1 above)
Comment (yhl25) [reply pkg/shared/clients/nats/nats_client.go]: reduced the ping interval to 3 seconds..

**Review by vigith — APPROVED** [review]

**Pull request merged**